File: test/goldens/charts/period_chart_golden_test.dart
@Tags(['golden'])
library;

import 'package:flutter_test/flutter_test.dart';

import 'package:openvitals/core/period/time_range.dart';
import 'package:openvitals/core/presentation/unit_formatter.dart';
import 'package:openvitals/core/time/local_date.dart';
import 'package:openvitals/domain/preferences/unit_system.dart';
import 'package:openvitals/ui/charts/period_chart.dart';
import 'package:openvitals/ui/theme/app_colors.dart';

import '../../support/golden_harness.dart';

/// [MetricBarChart] / [PeriodHistoryChart] — the most-used chart in the app, and
/// the only one that draws three genuinely different pictures from one call.
///
/// WEEK is bars, MONTH is a calendar heatmap, YEAR is a dot heatmap, and the
/// screen (sleep, mindfulness, hydration, steps…) just hands over the same list
/// of dated values and a range. So the range dispatch is what these shoot: three
/// ranges, one fixture, and the year one is the only place `yearAggregation`
/// shows up at all — a year of NIGHTS averages, because summing 365 sleeps would
/// print a number of hours nobody has ever slept.
void main() {
  final formatter = UnitFormatter(
    unitSystemProvider: () => UnitSystem.metric,
  );

  /// Hours slept per night, deterministically shaped: a long weekend lie-in, a
  /// short Wednesday, and the occasional night with no record at all (Health
  /// Connect gaps are the norm, not the exception).
  double sleepHours(LocalDate date) {
    if (date.day % 11 == 0) return 0; // no record that night
    final base = date.dayOfWeek >= 6 ? 8.4 : 7.0;
    return base + ((date.day * 3) % 7) * 0.12;
  }

  List<PeriodChartValue> nights(LocalDate start, LocalDate end) {
    final values = <PeriodChartValue>[];
    var date = start;
    while (!date.isAfter(end)) {
      final hours = sleepHours(date);
      if (hours > 0) values.add(PeriodChartValue(date, hours));
      date = date.plusDays(1);
    }
    return values;
  }

  testWidgets('week — bars', (tester) async {
    const week = DatePeriod(LocalDate(2026, 6, 16), LocalDate(2026, 6, 22));
    await expectChartGoldenBothThemes(
      tester,
      () => MetricBarChart(
        title: 'Sleep',
        values: nights(week.start, week.end),
        selectedRange: TimeRange.week,
        period: week,
        accentColor: AppColors.sleep,
        // Sleep's own alpha, not the 0.85 default — the sleep screen is the
        // caller that discovered this knob and the only one that moves it.
        accentAlpha: 0.75,
        summaryValue: '7h 32m avg',
        valueFormatter: (value) => '${formatter.decimal(value, 1)}h',
      ),
      name: 'period_chart_week',
    );
  });

  testWidgets('week — bars, with the selected day highlighted', (tester) async {
    const week = DatePeriod(LocalDate(2026, 6, 16), LocalDate(2026, 6, 22));
    await expectChartGoldenBothThemes(
      tester,
      () => MetricBarChart(
        title: 'Sleep',
        values: nights(week.start, week.end),
        selectedRange: TimeRange.week,
        period: week,
        accentColor: AppColors.sleep,
        accentAlpha: 0.75,
        summaryValue: '7h 32m avg',
        selectedDate: const LocalDate(2026, 6, 19),
        onDateSelected: (_) {},
        valueFormatter: (value) => '${formatter.decimal(value, 1)}h',
      ),
      name: 'period_chart_week_selected',
    );
  });

  testWidgets('month — the calendar heatmap', (tester) async {
    // The rolling 30-day window the period navigator actually hands over, hence
    // `last7Days`: the summary line has to say "Last 30 days", not "This month",
    // or it disagrees with the header above it.
    const month = DatePeriod(LocalDate(2026, 5, 24), LocalDate(2026, 6, 22));
    await expectChartGoldenBothThemes(
      tester,
      () => MetricBarChart(
        title: 'Sleep',
        values: nights(month.start, month.end),
        selectedRange: TimeRange.month,
        period: month,
        accentColor: AppColors.sleep,
        accentAlpha: 0.75,
        summaryValue: '7h 28m avg',
        weekPeriodMode: WeekPeriodMode.last7Days,
        valueFormatter: (value) => '${formatter.decimal(value, 1)}h',
      ),
      name: 'period_chart_month',
    );
  });

  testWidgets('year — the dot heatmap, twelve averaged months', (tester) async {
    const year = DatePeriod(LocalDate(2025, 7, 1), LocalDate(2026, 6, 22));
    await expectChartGoldenBothThemes(
      tester,
      () => MetricBarChart(
        title: 'Sleep',
        values: nights(year.start, year.end),
        selectedRange: TimeRange.year,
        period: year,
        accentColor: AppColors.sleep,
        accentAlpha: 0.75,
        summaryValue: '7h 30m avg',
        weekPeriodMode: WeekPeriodMode.last7Days,
        // Threaded through exactly as the sleep screen threads it (nights
        // average, they do not sum). The YEAR range dispatches to the heatmap,
        // which colours raw days, so this only bites on the bar path — it is
        // here because the caller sets it, not because the picture needs it.
        yearAggregation: PeriodBarAggregation.averageNonZero,
        valueFormatter: (value) => '${formatter.decimal(value, 1)}h',
      ),
      name: 'period_chart_year',
    );
  });
}

File: app/src/androidTest/kotlin/tech/mmarca/openvitals/ui/components/PeriodChartGoldenTest.kt
// The chart composables live under ui/charts/ but declare ui.components; this file mirrors that.
package tech.mmarca.openvitals.ui.components

import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Modifier
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.unit.dp
import java.time.LocalDate
import org.junit.Rule
import org.junit.Test
import tech.mmarca.openvitals.core.period.DatePeriod
import tech.mmarca.openvitals.core.period.TimeRange
import tech.mmarca.openvitals.core.period.WeekPeriodMode
import tech.mmarca.openvitals.core.presentation.DateTimeFormatterProvider
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import tech.mmarca.openvitals.domain.preferences.UnitSystem
import tech.mmarca.openvitals.testing.OpenVitalsVisualTestSurface
import tech.mmarca.openvitals.testing.assertVisualRootMatchesGolden
import tech.mmarca.openvitals.testing.goldenDates
import tech.mmarca.openvitals.ui.theme.SleepColor

/**
 * [MetricBarChart] / [PeriodHistoryChart]: WEEK is bars, MONTH a calendar heatmap, YEAR a dot
 * heatmap, from one call. The range dispatch is what these shoot.
 */
class PeriodChartGoldenTest {

    @get:Rule
    val composeRule = createComposeRule()

    @Test
    fun week_bars() {
        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 300.dp) {
                SleepBarChart(range = TimeRange.WEEK, period = WEEK, summary = "7h 32m avg")
            }
        }

        composeRule.assertVisualRootMatchesGolden("period_chart_week")
    }

    @Test
    fun week_barsWithTheSelectedDayHighlighted() {
        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 300.dp) {
                SleepBarChart(
                    range = TimeRange.WEEK,
                    period = WEEK,
                    summary = "7h 32m avg",
                    selectedDate = LocalDate.of(2026, 6, 19),
                    onDateSelected = {},
                )
            }
        }

        composeRule.assertVisualRootMatchesGolden("period_chart_week_selected")
    }

    @Test
    fun month_theCalendarHeatmap() {
        // The rolling 30-day window the navigator hands over, read off [LocalPeriodWeekMode].
        val month = DatePeriod(LocalDate.of(2026, 5, 24), LocalDate.of(2026, 6, 22))

        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 460.dp) {
                CompositionLocalProvider(LocalPeriodWeekMode provides WeekPeriodMode.LAST_7_DAYS) {
                    SleepBarChart(range = TimeRange.MONTH, period = month, summary = "7h 28m avg")
                }
            }
        }

        composeRule.assertVisualRootMatchesGolden("period_chart_month")
    }

    @Test
    fun year_theDotHeatmap() {
        // The year heatmap always draws the calendar year of `period.start`; that divergence is pinned.
        val year = DatePeriod(LocalDate.of(2025, 7, 1), LocalDate.of(2026, 6, 22))

        composeRule.setContent {
            OpenVitalsVisualTestSurface(width = 360.dp, height = 440.dp) {
                SleepBarChart(
                    range = TimeRange.YEAR,
                    period = year,
                    summary = "7h 30m avg",
                    yearAggregation = PeriodBarAggregation.AVERAGE_NON_ZERO,
                )
            }
        }

        composeRule.assertVisualRootMatchesGolden("period_chart_year")
    }

    @Composable
    private fun SleepBarChart(
        range: TimeRange,
        period: DatePeriod,
        summary: String,
        selectedDate: LocalDate? = null,
        onDateSelected: ((LocalDate) -> Unit)? = null,
        yearAggregation: PeriodBarAggregation = PeriodBarAggregation.SUM,
    ) {
        MetricBarChart(
            title = "Sleep",
            values = nights(period.start, period.end),
            selectedRange = range,
            period = period,
            accentColor = SleepColor,
            // Sleep's own alpha, not the 0.85 default.
            accentAlpha = 0.75f,
            summaryValue = summary,
            dateTimeFormatterProvider = DateTimeFormatterProvider(),
            yearAggregation = yearAggregation,
            selectedDate = selectedDate,
            onDateSelected = onDateSelected,
            valueFormatter = { "${FORMATTER.decimal(it, 1)}h" },
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
        )
    }

    private companion object {
        val FORMATTER = UnitFormatter(unitSystemProvider = { UnitSystem.METRIC })
        val WEEK = DatePeriod(LocalDate.of(2026, 6, 16), LocalDate.of(2026, 6, 22))

        /** Hours slept per night: a weekend lie-in, a short Wednesday, and nights with no record. */
        fun sleepHours(date: LocalDate): Double {
            if (date.dayOfMonth % 11 == 0) return 0.0
            val base = if (date.dayOfWeek.value >= 6) 8.4 else 7.0
            return base + ((date.dayOfMonth * 3) % 7) * 0.12
        }

        fun nights(start: LocalDate, end: LocalDate): List<PeriodChartValue> =
            goldenDates(start, end)
                .map { it to sleepHours(it) }
                .filter { (_, hours) -> hours > 0.0 }
                .map { (date, hours) -> PeriodChartValue(date, hours) }
    }
}
